src/main/java/graphql/execution/instrumentation/dataloader/DataLoaderCF.java

```diff
@@ -0,0 +1,99 @@
+package graphql.execution.instrumentation.dataloader;
+
+import graphql.ExperimentalApi;
+import graphql.Internal;
+import graphql.execution.DataLoaderDispatchStrategy;
+import graphql.execution.ExecutionContext;
+import graphql.schema.DataFetchingEnvironment;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
+import java.util.function.Supplier;
+
+import static graphql.execution.Execution.EXECUTION_CONTEXT_KEY;
+
+@Internal
+public class DataLoaderCF<T> extends CompletableFuture<T> {
+    final DataFetchingEnvironment dfe;
+    final String dataLoaderName;
+    final Object key;
+    final CompletableFuture<Object> dataLoaderCF;
+
+    volatile CountDownLatch latch;
+
+    public DataLoaderCF(DataFetchingEnvironment dfe, String dataLoaderName, Object key) {
+        this.dfe = dfe;
+        this.dataLoaderName = dataLoaderName;
+        this.key = key;
+        if (dataLoaderName != null) {
+            dataLoaderCF = dfe.getDataLoaderRegistry().getDataLoader(dataLoaderName).load(key);
+            dataLoaderCF.whenComplete((value, throwable) -> {
+                System.out.println("underlying DataLoader completed");
+                if (throwable != null) {
+                    completeExceptionally(throwable);
+                } else {
+                    complete((T) value);
+                }
+                // post completion hook
+                if (latch != null) {
+                    latch.countDown();
+                }
+            });
+        } else {
+            dataLoaderCF = null;
+        }
+    }
+
+    DataLoaderCF() {
+        this.dfe = null;
+        this.dataLoaderName = null;
+        this.key = null;
+        dataLoaderCF = null;
+    }
+
+    @Override
+    public <U> CompletableFuture<U> newIncompleteFuture() {
+        return new DataLoaderCF<>();
+    }
+
+    public static boolean isDataLoaderCF(Object object) {
+        return object instanceof DataLoaderCF;
+    }
+
+    @ExperimentalApi
+    public static <T> CompletableFuture<T> newDataLoaderCF(DataFetchingEnvironment dfe, String dataLoaderName, Object key) {
```

Comment:
In a future version this wont be static but rather come from some "ChainedDLFactory" given out to DataFetchers

Therefore the lookup to get a `DataLoader` can be done in that method

```
DataFetcher df = dfe -> {
   return dfe.getSomeFactory(). newDataLoaderCF(actualDataLoader, key).thenApply(x -> x.key)
}
```

```